### PR TITLE
Add cache timings to traces

### PR DESCRIPTION
**Pull request by saolsen**
Rough draft of tracing cache hits.

There are a few problems with it though. The only way I could figure out how to pass in the span is at the root when the data-loaders are constructed so the span is parented to the query instead of it's individual resolver. You can see in the image that it's still shown in the right part chronologically but the tree is broken so you can't see all the cache hits for a single resolver, just for the whole query.

It also feels kind of hacky to do it this way so if you have other ideas I'd like to hear them.

Also there's the fact that we need to replace the datadog library we're using soon too so is it worth adding this now or waiting until after that.

If we just want another quick vis tool for debugging right now this could be enough.

chart

### CODE REVIEW

**Comment by acjay** [comment]
Spent some time thinking about this, as you can see below. But then, I came to realize that resolver spans probably don't really _own_ the actual cache/API requests, since there can be a one-to-many mapping from requesting a key to the loading of that key. Maybe it would make sense to have spans for each data loader instance, which would be the parent the spans for individual key fetches? And then on the resolver side, a span would be created for the call to `load`? (Although that last part probably would require some dynamic wrapping of `load`, as alluded to below.)

## Old thought

Could you use a similar trick as you did for tracing the resolvers? If you wrapped the data loaders in `wrapResolve`, you could replace them with decorated versions that start and finish the spans. You could use something like [a proxy object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) to only do this on-demand when a loader is actually accessed by a resolver.

Something like:

```javascript
return function wrappedResolver(_root, _opts, _req, _info) {
  const { rootValue } = _info
  // ...
  const rootValueHandler = {
    get(rv, property) {
      // intercept calls to data loaders in question  
    }
  }
  const rootValueProxy = new Proxy(rootValue, rootValueHandler)
  const result = resolver.call(this, _root, _opts, _req, { ..._info, rootValue: rootValueProxy })
```

**Comment by anandaroop** [comment]
Since we're not rushing to merge this, I'm adding a [link](https://trello.com/c/CwFpcJIg/101-add-mp-cache-info-to-datadog) to the #incidents card here — possible follow-ups, whether in this PR or another one.

**Comment by alloy** [comment]
> I came to realize that resolver spans probably don't really own the actual cache/API requests, since there can be a one-to-many mapping from requesting a key to the loading of that key. Maybe it would make sense to have spans for each data loader instance, which would be the parent the spans for individual key fetches?

Indeed, many resolvers could invoke a data loader and the data loader instance could then decide to invoke a single fetch, so I do agree with this suggestion; however, it would still be nice if there was a trace of which resolvers invoked the data loader instance. Is it possible to associate multiple parent spans with a single child span?

**Comment by saolsen** [comment]
Nah I think you can only have one parent. The utility of the parenting is only in how datadog uses it to lay out that graph so it's only important to get it right enough that the view is useful. I think since it's not 1:1 with resolvers then leaving them children of the main query is fine.

This just traces the cache get but I think it would make sense to trace all the dataloader fetches too. I'll add that and see how it looks. I think It would show refreshes that happen after the main request is finished too which would give us a better picture of how metaphysics actually works.

Also now that we're not rushing this as a debugging diagnostic I'll look into getting on the right tracing library (and we still need to follow up with datadog on why distributed tracing is broken)

**Comment by alloy** [comment]
Sounds good 👍 One note is that the work I’m doing to add timeouts to resolvers is using https://github.com/prismagraphql/graphql-middleware/ so it would be nice to consolidate on that in favor of the manual resolver wrapping that we’re doing atm for tracing purposes.

**Comment by mzikherman** [comment]
@saolsen thanks for this! Checking it out now.

I'd be down to merge anything that's still a WIP, as long as it's better than what we currently have. There's still some commented out code here though, I'm not sure if this is truly mergeable or those should be un-commented? Or are those the things that aren't working and need to be addressed?

> The individual resolver tracing and the dataloader tracing still needs to be done but I have to drop this for now (though if anyone wants to pick it up they're welcome to.)

I'm down to finish this, but we could/should probably prioritize finishing this up and getting monitoring back in a good state over regular sprint work, IMO.

Maybe we can sync up quickly on the remaining items and I'll finish it up? I'm on-call this week so can definitely take some time to do this.

Thanks again for the push here! 🍾

**Comment by saolsen** [comment]
@mzikherman I added back in the resolver tracing and got rid of that commented out half done stuff so this should be ready for final review and merging.

**Comment by mzikherman** [comment]
Sweet! 😻 

Will merge and check it out on staging.

File: src/lib/tracer.js [hunk 1]
```diff
@@ -29,42 +24,18 @@ function drop_params(query) {
...
-function trace(res, span) {
-  span.setTag("http.status_code", res.statusCode)
+export async function traceMiddleware(resolve, parent, args, ctx, info) {
+  const span = await tracer.trace("graphql.resolver", {
+    resource: info.parentType + ": " + info.fieldName,
+  })
+  const result = await resolve(parent, args, ctx, info)
   span.finish()
+  return result
 }
```
Comment (alloy) [inline src/lib/tracer.js]: Nice 👌

File: src/lib/tracer.js [hunk 2]
```diff
@@ -19,84 +24,37 @@ function drop_params(query) {
...
-      forIn(type._fields, (field, fieldName) => {
-        if (
-          field.resolve instanceof Function &&
-          getNamedType(field.type) instanceof GraphQLObjectType
-        ) {
-          field.resolve = wrapResolve(typeName, fieldName, field.resolve) // eslint-disable-line no-param-reassign
-        }
-      })
-    }
-  })
```
Comment (alloy) [inline src/lib/tracer.js]: 👋

**Pull request merged**